### PR TITLE
invite-accept: drop 'fetchInvite' flux action

### PR DESCRIPTION
**Pull request by flootr**
This is the first PR in a series which removes the remaining `lib/invites` flux stores.

## Changes proposed in this Pull Request

* drop `fetchInvite` flux action
* remove `invite-accept-validation` flux store
* use component state exclusively

Note: I chose using internal component state because data is used exclusively in that component, nothing else is subscribing to it.

#### Testing instructions

Go to `/people/invites/<site>` and create a new invitation (the details don't matter that much) for a test account of yours. Hit the link you should receive via email. Replace the `wordpress.com` part of the URL with the the calypso.live link from the comment below.

The form should look more or less like the screenshot below. This PR only touches the part where we fetch invite data so if you see something like that, this should be fine. If you're not logged in you should see a signup form. In no case you should see an error.

Screenshot 2020-11-17 at 15 55 53

Now manipulate the invitation key in the URL and make sure you see an error like this:

Screenshot 2020-11-17 at 15 58 22

part of #24149

### CODE REVIEW

File: client/my-sites/invites/invite-accept/index.jsx [hunk 1]
```diff
@@ -49,32 +49,34 @@ class InviteAccept extends React.Component {
...
 	}
 
-	refreshInvite = () => {
-		const invite = InvitesStore.getInvite( this.props.siteId, this.props.inviteKey );
-		const error = InvitesStore.getInviteError( this.props.siteId, this.props.inviteKey );
+	async fetchInvite() {
+		try {
+			const response = await wpcom
+				.undocumented()
+				.getInvite( this.props.siteId, this.props.inviteKey );
```
Comment (tyxla) [inline client/my-sites/invites/invite-accept/index.jsx:71]: Would be great if we could use the data layer for that one. Can be done in another PR though.

File: client/my-sites/invites/invite-accept/utils/normalize-invite.js [hunk 2]
```diff
@@ -40,19 +33,3 @@ function normalizeInvite( data ) {
 		knownUser: data.invite.meta.known,
 	};
 }
-
-const reducer = ( state = initialState, payload ) => {
-	const { action } = payload;
-	switch ( action.type ) {
-		case ActionTypes.RECEIVE_INVITE:
```
Comment (tyxla) [inline client/my-sites/invites/invite-accept/utils/normalize-invite.js:47]: This `RECEIVE_INVITE` constant from `'calypso/lib/invites/constants'` will no longer be used if we merge this PR. Can we completely delete it as well?

File: client/my-sites/invites/invite-accept/utils/normalize-invite.js [hunk 3]
```diff
@@ -40,19 +33,3 @@ function normalizeInvite( data ) {
...
-			return state.setIn(
-				[ 'list', action.siteId, action.inviteKey ],
-				normalizeInvite( action.data )
-			);
-		case ActionTypes.RECEIVE_INVITE_ERROR:
```
Comment (tyxla) [inline client/my-sites/invites/invite-accept/utils/normalize-invite.js:52]: This `RECEIVE_INVITE_ERROR ` constant from `'calypso/lib/invites/constants'` will no longer be used if we merge this PR. Can we completely delete it as well?

File: client/my-sites/invites/invite-accept/utils/normalize-invite.js [hunk 4]
```diff
@@ -26,7 +19,7 @@ function filterObjectProperties( object ) {
 	} );
 }
 
-function normalizeInvite( data ) {
+export default function normalizeInvite( data ) {
```
Comment (tyxla) [inline client/my-sites/invites/invite-accept/utils/normalize-invite.js:22]: This normalization would be a great fit to be ported to a `fromApi` transformer once we port the wpcom request to the data layer.

File: client/lib/invites/actions.js [hunk 5]
```diff
@@ -23,32 +23,6 @@ import { requestSiteInvites } from 'calypso/state/invites/actions';
  */
 const debug = new Debug( 'calypso:invites-actions' );
 
-export function fetchInvite( siteId, inviteKey ) {
-	debug( 'fetchInvite', siteId, inviteKey );
-
-	Dispatcher.handleViewAction( {
-		type: ActionTypes.FETCH_INVITE,
```
Comment (tyxla) [inline client/lib/invites/actions.js:30]: This `FETCH_INVITE ` constant from `'calypso/lib/invites/constants'` will no longer be used if we merge this PR. Can we completely delete it as well?

File: client/my-sites/invites/invite-accept/index.jsx [hunk 6]
```diff
@@ -49,32 +49,34 @@ class InviteAccept extends React.Component {
...
+				...normalizeInvite( response ),
 				activationKey: this.props.activationKey,
 				authKey: this.props.authKey,
+			};
+
+			this.setState( { invite } );
+		} catch ( error ) {
+			this.setState( { error } );
+
+			recordTracksEvent( 'calypso_invite_validation_failure', {
```
Comment (tyxla) [inline client/my-sites/invites/invite-accept/index.jsx:83]: Would be nice if we could port this to use the Redux `recordTracksEvent` instead, too. Not a requirement for this PR though; can be done in another one.

File: client/my-sites/invites/invite-accept/index.jsx [hunk 7]
```diff
@@ -49,32 +49,34 @@ class InviteAccept extends React.Component {
 		matchEmailError: false,
 	};
 
-	UNSAFE_componentWillMount() {
+	componentDidMount() {
 		// The site ID and invite key are required, so only fetch if set
 		if ( this.props.siteId && this.props.inviteKey ) {
-			fetchInvite( this.props.siteId, this.props.inviteKey );
+			this.fetchInvite();
```
Comment (tyxla) [inline client/my-sites/invites/invite-accept/index.jsx:59]: A couple of questions here:

* Should we abort the request if the user navigates away?
* If this component is mounted twice in a relatively short period of time, will the request be doubled? This is related to my comment about migrating to the data layer, as that'd solve such a potential problem.

**Review by tyxla — APPROVED** [review]
Nice work! This is testing great 👍  

Also, code looks good. I've left some nits and suggestions, but nothing blocking in my opinion. 

Thanks for using component state, this is a great example of a case where redux is just unnecessary. 💯

File: client/my-sites/invites/invite-accept/index.jsx (hunk 1 above)
Comment (flootr) [reply client/my-sites/invites/invite-accept/index.jsx:71]: I'm a little bit confused. I'm not against using the data layer but that would mean leveraging redux, no?

File: client/my-sites/invites/invite-accept/index.jsx (hunk 7 above)
Comment (jsnajdr) [reply client/my-sites/invites/invite-accept/index.jsx:59]: @tyxla The `invites-accept-validation` store is used only in this component, and stores only the response from one `fetchInvite` request. The invite object ceases to exist once it's successfully accepted, so it doesn't need to be cached and stored by `siteId` and `inviteKey`.

That's probably why @flootr decided to make it a local state of the component. Data layer would require a Redux store for the state.

> Should we abort the request if the user navigates away?

That's valid -- there should be an `.isMounted` check before calling `.setState` in the promise callbacks.

> If this component is mounted twice in a relatively short period of time, will the request be doubled?

That shouldn't be a concern for a request that only reads stuff and doesn't do modifications.

The data layer deduplication is useful mainly if there are two independent components on a page that query or use the same data. Then it makes sure the requests are co-ordinated and the responses are shared.

Comment (flootr) [reply client/my-sites/invites/invite-accept/index.jsx:59]: Thanks for chiming in @jsnajdr. Would d8bebe1 work in that regard?

File: client/my-sites/invites/invite-accept/index.jsx (hunk 1 above)
Comment (tyxla) [reply client/my-sites/invites/invite-accept/index.jsx:71]: Yeah. Let's skip this for now. I agree to keep this one simple - porting it to data layer will require some additional Redux work, and I like that this one is Redux-free.

**Review by jsnajdr — APPROVED** [review]
Looks good to me :+1:

**Pull request merged**